server/client.go

```diff
@@ -974,61 +1253,95 @@ func (c *client) deliverMsg(sub *subscription, mh, msg []byte) bool {
 	client.outMsgs++
 	client.outBytes += msgSize
 
-	atomic.AddInt64(&c.srv.outMsgs, 1)
-	atomic.AddInt64(&c.srv.outBytes, msgSize)
-
-	// Check to see if our writes will cause a flush
-	// in the underlying bufio. If so limit time we
-	// will wait for flush to complete.
+	atomic.AddInt64(&srv.outMsgs, 1)
+	atomic.AddInt64(&srv.outBytes, msgSize)
 
-	deadlineSet := false
-	if client.bw.Available() < (len(mh) + len(msg)) {
-		client.wfc++
-		client.nc.SetWriteDeadline(time.Now().Add(client.srv.getOpts().WriteDeadline))
-		deadlineSet = true
-	}
+	// Queue to outbound buffer
+	client.queueOutbound(mh)
+	client.queueOutbound(msg)
 
-	// Deliver to the client.
-	_, err := client.bw.Write(mh)
-	if err != nil {
-		goto writeErr
-	}
+	client.out.pm++
 
-	_, err = client.bw.Write(msg)
-	if err != nil {
-		goto writeErr
+	// Check outbound threshold and queue IO flush if needed.
+	if client.out.pm > 1 && client.out.pb > maxBufSize*2 {
+		client.flushSignal()
 	}
 
 	if c.trace {
 		client.traceOutOp(string(mh[:len(mh)-LEN_CR_LF]), nil)
 	}
 
-	// TODO(dlc) - Do we need this or can we just call always?
-	if deadlineSet {
-		client.nc.SetWriteDeadline(time.Time{})
+	// Increment the flush pending signals if we are setting for the first time.
+	if _, ok := c.pcd[client]; !ok {
+		client.out.fsp++
 	}
-
 	client.mu.Unlock()
+
+	// Remember for when we return to the top of the loop.
 	c.pcd[client] = needFlush
+
 	return true
+}
 
-writeErr:
-	if deadlineSet {
-		client.nc.SetWriteDeadline(time.Time{})
+// pruneCache will prune the cache via randomly
+// deleting items. Doing so pruneSize items at a time.
+func (c *client) prunePubPermsCache() {
+	r := 0
+	for subject := range c.perms.pcache {
+		delete(c.perms.pcache, subject)
+		if r++; r > pruneSize {
+			break
+		}
+	}
+}
+
+// pubAllowed checks on publish permissioning.
+func (c *client) pubAllowed() bool {
+	// Disallow publish to _SYS.>, these are reserved for internals.
+	if c.pa.subject[0] == '_' && len(c.pa.subject) > 4 &&
```

Comment:
I know that this is moved and not new code, but someone previously reported that bytes.Equal() would probably be as fast if not faster?

Reply:
Used to be much slower but you are probably right. I could run a micro benchmark or just check the normal benchmarks. Will test quickly and if so will change.

Reply:
Will need to be bytes.HasPrefix(). Running benchmarks now.

Reply:
5ns slower with bytes.HasPrefix(), so will leave as is.

Reply:
Ok, I was basing my comments on [this](https://github.com/nats-io/gnatsd/pull/571#pullrequestreview-92844128).

Reply:
but agreed, not worth the change.

Reply:
I tweaked a bit and it made a small difference with string cast and subslice with ==.